synapse/util/distributor.py

```diff
@@ -128,31 +97,21 @@ def fire(self, *args, **kwargs):
         """Invokes every callable in the observer list, passing in the args and
         kwargs. Exceptions thrown by observers are logged but ignored. It is
         not an error to fire a signal with no observers.
+        """
 
-        Returns a Deferred that will complete when all the observers have
-        completed."""
-
-        def do(observer):
-            def eb(failure):
+        async def do(observer):
+            try:
+                result = observer(*args, **kwargs)
+                if inspect.isawaitable(result):
+                    result = await result
+                return result
+            except Exception as e:
                 logger.warning(
-                    "%s signal observer %s failed: %r",
-                    self.name,
-                    observer,
-                    failure,
-                    exc_info=(
-                        failure.type,
-                        failure.value,
-                        failure.getTracebackObject(),
-                    ),
+                    "%s signal observer %s failed: %r", self.name, observer, e,
                 )
 
-            return maybeAwaitableDeferred(observer, *args, **kwargs).addErrback(eb)
-
-        deferreds = [run_in_background(do, o) for o in self.observers]
-
-        return make_deferred_yieldable(
-            defer.gatherResults(deferreds, consumeErrors=True)
-        )
+        for observer in self.observers:
+            run_in_background(do, observer)
```

Comment:
this is changing the semantics so that we no longer wait for the observers to complete. Is that deliberate? what is the logic behind it?

Reply:
The reasoning behind this nothing uses the `Deferred` returned by `Signal.fire()`. [`gatherResults`](https://twistedmatrix.com/documents/current/api/twisted.internet.defer.html#gatherResults) and `make_deferred_yieldable` don't seem to actually "wait" since the caller doesn't wait on the result.

I can back out 6c27891edc0b2a5747b3501cf88303ae262da79a if you'd like.

Reply:
https://github.com/matrix-org/synapse/pull/8216/files/b400db3b5bd2d365d4e68b73f7d3f2a0530dd8a7#diff-755d6db2f812691f4b7ea7fd62894893R72 calls Signal.fire (and uses the deferred, via run_as_background_process) ?

> I can back out 6c27891 if you'd like.

I think that would be correct.

Reply:
Ah, good call. 👍 I guess I was looking at places that `Distributor.fire` is called. 😢

Reply:
Reverted in 41ced4c7cf4b67d0ccd05c271be170bc9fc51253